x-pack/apm-server/aggregation/spanmetrics/aggregator.go

```diff
@@ -0,0 +1,243 @@
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+		config:   config,
+		active:   newMetricsBuffer(),
+		inactive: newMetricsBuffer(),
+	}, nil
+}
+
+func (a Aggregator) Run() error {
+	ticker := time.NewTicker(a.config.Interval)
+	defer ticker.Stop()
+	defer func() {
+		a.stopMu.Lock()
+		defer a.stopMu.Unlock()
+		select {
+		case <-a.stopped:
+		default:
+			close(a.stopped)
+		}
+	}()
+	var stop bool
+	for !stop {
+		select {
+		case <-a.stopping:
+			stop = true
+		case <-ticker.C:
+		}
+		if err := a.publish(context.Background()); err != nil {
+			a.config.Logger.With(logp.Error(err)).Warnf(
+				"publishing span metrics failed: %s", err,
+			)
+		}
+	}
+	return nil
+}
+
+// Stop stops the Aggregator if it is running, waiting for it to flush any
+// aggregated metrics and return, or for the context to be cancelled.
+//
+// After Stop has been called the aggregator cannot be reused, as the Run
+// method will always return immediately.
+func (a *Aggregator) Stop(ctx context.Context) error {
+	a.stopMu.Lock()
+	select {
+	case <-a.stopped:
+	case <-a.stopping:
+		// Already stopping/stopped.
+	default:
+		close(a.stopping)
+	}
+	a.stopMu.Unlock()
+
+	select {
+	case <-a.stopped:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	return nil
+}
+
+func (a *Aggregator) publish(ctx context.Context) error {
+	// We hold a.mu only long enough to swap the spanMetrics. This will
+	// be blocked by spanMetrics updates, which is OK, as we prefer not
+	// to block spanMetrics updaters. After the lock is released nothing
+	// will be accessing a.inactive.
+	a.mu.Lock()
+	a.active, a.inactive = a.inactive, a.active
+	a.mu.Unlock()
+
+	size := a.inactive.size()
+	if size == 0 {
+		a.config.Logger.Debugf("no span metrics to publish")
+		return nil
+	}
+
+	now := time.Now()
+	metricsets := make([]transform.Transformable, 0, size)
+	for hash, metrics := range a.inactive.m {
+		metricset := makeMetricset(now, metrics.key, metrics.count, metrics.sum)
+		metricsets = append(metricsets, &metricset)
+		delete(a.inactive.m, hash)
+	}
+
+	a.config.Logger.Debugf("publishing %d metricsets", len(metricsets))
+	return a.config.Report(ctx, publish.PendingReq{
+		Transformables: metricsets,
+		Trace:          true,
+	})
+}
+
+func (a *Aggregator) AggregateTransformables(in []transform.Transformable) {
+	for _, tf := range in {
+		if span, ok := tf.(*model.Span); ok &&
+			span.DestinationService != nil &&
+			span.DestinationService.Resource != nil {
+			key := aggregationKey{
+				serviceEnvironment: span.Metadata.Service.Environment,
+				serviceName:        span.Metadata.Service.Name,
+				resource:           *span.DestinationService.Resource,
+			}
+			duration := time.Duration(span.Duration * float64(time.Millisecond))
+			metrics := spanMetrics{
+				count: span.RepresentativeCount,
+				sum:   duration.Milliseconds(),
+				key:   key,
+			}
+			a.active.storeOrUpdate(metrics)
+		}
+	}
+}
+
+type metricsBuffer struct {
+	// TODO might need a size cap
```

Comment:
Can we measure that before taking that decision? Eg. what workload is needed with the default interval for this to have a non-negligible impact? Does that workload even exist?

Reply:
We could do a deep analysis, but I don't think we should spend time on that right now.

A very cursory analysis:
 - RUM is typically unauthenticated
 - The default event rate limit is 300 per IP
 - The default number of concurrent IPs for which we track event rate limits is 1000

So a malicious actor with access to 1000 unique IPs (botnet) could send 300000 unique service names/destination resources per second. That's already pretty high, but if the rate limits are increased, so is the risk.

Reply:
A single APM Server can not ingest 300k events / second, I don't think it is even close to realistic. I think you would need already a big setup to reach 2k/sec, correct me if I am wrong.
So worst realistic-ish case would be to end up with a map with 100-300k keys. How much memory is that map expected to need? 
How much memory will be consuming the server on the other hand to validate and index 2k events / sec? What is the ratio?

> if the rate limits are increased, so is the risk

That is the user's choice, they have the tools to adjust it.
Does it look all that bad (for an extreme case like a DDoS attack to an apm-server) to warrant the extra complexity and development and maintenance time, already in the first iteration, and without measuring?

Reply:
> A single APM Server can not ingest 300k events / second, I don't think it is even close to realistic. I think you would need already a big setup to reach 2k/sec, correct me if I am wrong.

Yes, 300k/sec is an upper bound.

Right now we won't get close. In the not too distant future, we should be able to do significantly more than today with tail-sampling (avoids indexing, which means less back pressure), and the decoder improvements (speeds up decoding of course).

>> if the rate limits are increased, so is the risk
>
> That is the user's choice, they have the tools to adjust it.

Sure, but I'd like to avoid pushing the problem onto the user if we can.

> Does it look all that bad (for an extreme case like a DDoS attack to an apm-server) to warrant the extra complexity and development and maintenance time, already in the first iteration, and without measuring?

I'm struggling to understand your objection about having a limit. It seems you think it's much more difficult to implement and maintain than what I expect. Let's discuss sync to figure it out.

Reply:
Sure!

Reply:
> Right now we won't get close. In the not too distant future, we should be able to do significantly more than today with tail-sampling (avoids indexing, which means less back pressure), and the decoder improvements (speeds up decoding of course).

I just ran a test on my laptop: I can push 3k/sec with the master branch, and with my tail-sampling branch I hit 50k/sec.

To be fair these are mostly empty transactions, but I don't think it's outside the realms of possibility that we would hit memory limits in smaller cloud instances. Particularly once decoding/validation is optimised, and also when the tail-sampling code is more optimised.